test/gtest/uct/test_pending.cc

```diff
@@ -254,6 +281,53 @@ UCS_TEST_P(test_uct_pending, send_ooo_with_pending)
     EXPECT_EQ(exp_counter, counter);
 }
 
+/*
+ * test that the pending op callback is only called from the progress()
+ */
+UCS_TEST_P(test_uct_pending, pending_async)
+{
+    pending_send_request_t *req = NULL;
+    ucs_status_t status;
+    size_t packed_len;
+
+    initialize();
+    check_caps(UCT_IFACE_FLAG_AM_BCOPY | UCT_IFACE_FLAG_PENDING);
+
+    mapped_buffer sbuf(ucs_min(64ul, m_e1->iface_attr().cap.am.max_bcopy), 0,
+                       *m_e1);
+
+    req = pending_alloc_simple(&sbuf, 0);
+
+    /* set a callback for the uct to invoke when receiving the data */
+    install_handler_sync_or_async(m_e2->iface(), 0, am_handler_simple, 0);
+
+    /* send while resources are available */
+    n_pending = 0;
+    do {
+        packed_len = uct_ep_am_bcopy(m_e1->ep(0), 0, mapped_buffer::pack,
+                                     &sbuf, 0);
+    } while ((int)packed_len >= 0);
+
+    EXPECT_TRUE((int)packed_len == UCS_ERR_NO_RESOURCE);
+
+    status = uct_ep_pending_add(m_e1->ep(0), &req->uct);
+    EXPECT_UCS_OK(status);
+    n_pending++;
+
+    /* pending op must not be called either asynchronously or from the
+     * uct_ep_am_bcopy() */
+    twait(300);
+    EXPECT_EQ(1, n_pending);
+
+    packed_len = uct_ep_am_bcopy(m_e1->ep(0), 0, mapped_buffer::pack, &sbuf, 0);
+    EXPECT_EQ(1, n_pending);
+    EXPECT_LT(0, (int)packed_len);
+
+    short_progress_loop();
```

Comment:
with high load one loop may not be enough to progress all outstanding sends + pending. Better use `wait_for_flag` for n_pending or similar